### PR TITLE
Guidpcg

### PR DESCRIPTION
**Pull request by khlebnikov**

### CODE REVIEW

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 1]
```diff
@@ -20,34 +25,73 @@ namespace bdlb {
...
+        *hex = 13;
+        break;
+      case 'e':
+      case 'E':
+        *hex = 14;
+        break;
+      case 'f':
+      case 'F':
+        *hex = 15;
+        break;
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]: Be careful when running `bde_format` on the entire file.  Unfortunately, some things are manually formatted - please revert this change.

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 2]
```diff
@@ -94,13 +137,30 @@ int vaildateGuidString(bslstl::StringRef guidString)
...
+          case 'a':
+          case 'b':
+          case 'c':
+          case 'd':
+          case 'e':
+          case 'f':
             ++digitCnt;
-          case '-':
-            ;
+          case '-':;
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]: Please revert

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 3]
```diff
@@ -137,6 +197,39 @@ void GuidUtil::generate(Guid *result, bsl::size_t numGuids)
     generate(reinterpret_cast<unsigned char *>(result), numGuids);
 }
 
+void GuidUtil::generateNonSecure(unsigned char *result, bsl::size_t numGuids)
+{
+    unsigned char *bytes = result;
+    unsigned char *end   = bytes + numGuids * Guid::k_GUID_NUM_BYTES;
+
+    static bdlb::PCG pcgSingleton;
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]: This is not a trivial constructor.  I suggest that we have a pointer outside the `ONCE_DO` block...

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 4]
```diff
@@ -137,6 +197,39 @@ void GuidUtil::generate(Guid *result, bsl::size_t numGuids)
...
+    BSLMT_ONCE_DO
+    {
+        uint64_t seed;
+        if (0 != RandomDevice::getRandomBytes((unsigned char *)&seed,
+                                              sizeof(seed))) {
+            seed = time(NULL) ^ (intptr_t)&bsl::printf;  // fallback seed
+        }
+        int rounds = 5;  // see PCG code base for details of 'rounds'
+        pcgSingleton.seed(seed, (intptr_t)&rounds);
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]: ... and here we construct a `static bdlb::PCG pcgSingleton(seed, streamSelector);` and then set the pointer. 

Also, please adjust this initialization because `(intptr_t)&rounds` doesn't make much sense. I'd expect the stream selector to also come from `getRandomBytes`.

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 5]
```diff
@@ -137,6 +197,39 @@ void GuidUtil::generate(Guid *result, bsl::size_t numGuids)
...
+    }
+    {
+        static bslmt::Mutex            mutex;
+        bslmt::LockGuard<bslmt::Mutex> guard(&mutex);
+        size_t                         numBytesToFill = end - bytes;
+        for (size_t i = 0; i < numBytesToFill; i += sizeof(bsl::uint32_t)) {
+            bsl::uint32_t rnd_int = pcgSingleton.generate();
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]: Please adjust the `rnd_int` variable name to meet our coding standard.

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 6]
```diff
@@ -199,6 +199,20 @@ struct GuidUtil {
...
+        // specification, and load the resulting GUIDs into the array referred
+        // to by the specified 'result'.  Optionally specify 'numGuids',
+        // indicating the number of GUIDs to load into the 'result' array.  If
+        // 'numGuids' is not supplied, a default of 1 is used.  An RFC 4122
+        // version 4 GUID consists of 122 randomly generated bits, two
+        // 'variant' bits set to '10' and four 'version' bits set to '0100'.
+        // The behavior is undefined unless 'result' refers to a contiguous
+        // sequence of at least 'numGuids' Guid objects.  Uses the PCG random
+        // engine to generate high quality random numbers for Guids.  This
+        // function does not generate cryptographically secure Guids.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]:
```suggestion
        // sequence of at least 'numGuids' Guid objects.  Note that this function uses the PCG random
        // engine to generate high quality, albeit not cryptographically secure, random numbers for Guids. 
```
(+bde_format)

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 7]
```diff
@@ -233,10 +247,7 @@ struct GuidUtil {
 //                      INLINE DEFINITIONS
 // ============================================================================
 
-                              // ---------------
-                              // struct GuidUtil
-                              // ---------------
-// CLASS METHODS
+// --------------- struct GuidUtil --------------- CLASS METHODS
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]: Revert

**Review by khlebnikov — COMMENTED** [review]
Added some comments to the GuidUtil integration.  Will make another review round when you apply all the changes we discussed.

File: groups/bdl/bdlb/bdlb_pcg.h [hunk 8]
```diff
@@ -83,10 +73,10 @@ class PCG {
...
     // CREATORS
     PCG(bsl::uint64_t initState = 0, bsl::uint64_t streamSelector = 1);
-        // Create a 'PCG' object and seed it with the specified 'initState' and
-        // 'streamSelector.' For the description of the parameters, refer to
-        // the 'seed' method. Note that 'seed' is invoked in the body of this
-        // constructor.
+        // Create a 'PCG' object and seed it with the optionally specified
+        // 'initState' and 'streamSelector.'  If 'initState' is not specified,
+        // 0 is used as the initial state.  If 'streamSelector' is not
+        // specified, 1 is used as the initial state.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcg.h]:
```suggestion
        // specified, 1 is used as the stream selector.
```

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 9]
```diff
@@ -132,6 +137,45 @@ void GuidUtil::generate(unsigned char *result, bsl::size_t numGuids)
...
+    unsigned char *end   = bytes + numGuids * Guid::k_GUID_NUM_BYTES;
+
+    static bdlb::PCGRandomGenerator *pcgSingletonPtr;
+    BSLMT_ONCE_DO
+    {
+        uint64_t state;
+        if (0 != RandomDevice::getRandomBytes((unsigned char *)&state,
+                                              sizeof(state))) {
+            state = time(NULL) ^ (intptr_t)&bsl::printf;  // fallback state
+        }
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp:170]: (for the future) This might be a candidate for factoring into local function - this is repeated twice here.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+//
+//
+///Usage
+///-----
+// This section illustrates intended use of this component.
+//
+///Example 1: Generating random 32-bit numbers
+///- - - - - - - - - - - - - - - - - - - - - -
+// The PCGRandomGenerator constructor takes two 64-bit constants (the initial
+// state, and the rng sequence selector.  Instances of PCGRandomGenerator with
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// state, and the rng sequence selector.  Instances of 'PCGRandomGenerator' with
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+// different sequence selectors will never have random sequences that coincide.
+// In order to obtain nondeterministic output for each instance, on a Unix
+// system random bytes can be obtained from /dev/urandom.  The component 
+// bdlb_randomdevice provides such functionality.  Also illustrated below is a 
+// a backup mechanism for obtaining random bytes.
+//
+//  uint64_t state;
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
//..
//  uint64_t state;
```

Comment (khlebnikov) [reply groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: We'll bring the wording to more BDE style later.  For now, just couple formatting nits.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+///-----
+// This section illustrates intended use of this component.
+//
+///Example 1: Generating random 32-bit numbers
+///- - - - - - - - - - - - - - - - - - - - - -
+// The PCGRandomGenerator constructor takes two 64-bit constants (the initial
+// state, and the rng sequence selector.  Instances of PCGRandomGenerator with
+// different sequence selectors will never have random sequences that coincide.
+// In order to obtain nondeterministic output for each instance, on a Unix
+// system random bytes can be obtained from /dev/urandom.  The component
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// system random bytes can be obtained from '/dev/urandom'.  The component 
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+//..
+
+#include <bsl_cstdint.h>
+
+namespace BloombergLP {
+namespace bdlb {
+
+                                 // =========
+                                 // class PCGRandomGenerator
+                                 // =========
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: bde-format this

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+// FREE OPERATORS
+bool operator==(const PCGRandomGenerator& lhs, const PCGRandomGenerator& rhs);
+    // Return 'true' if the specified 'lhs' and 'rhs' objects have the same
+    // value, and 'false' otherwise.  Two 'PCGRandomGenerator' objects have the same value if
+    // both of the corresponding values of their 'd_state' and
+    // 'd_streamSelector' attributes are the same.
+
+bool operator!=(const PCGRandomGenerator& lhs, const PCGRandomGenerator& rhs);
+    // Return 'true' if the specified 'lhs' and 'rhs' objects do not have the
+    // same value, and 'false' otherwise.  Two 'PCGRandomGenerator' objects do not have the
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: bde-format

File: groups/bdl/bdlb/bdlb_guidutil.cpp (hunk 9 above)
Comment (bottha123) [reply groups/bdl/bdlb/bdlb_guidutil.cpp:170]: Yes, I considered that. Also in the GuidUtil implementation, the following code is duplicated (by my new function). Perhaps a candidate for factoring as well:
    while (bytes != end) {
        typedef unsigned char uc;
        bytes[6] = uc(0x40 | (bytes[6] & 0x0F));
        bytes[8] = uc(0x80 | (bytes[8] & 0x3F));
        bytes += Guid::k_GUID_NUM_BYTES;
    }
My reasoning there was that I ought not touch working code.

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 15]
```diff
@@ -199,6 +199,20 @@ struct GuidUtil {
...
+        // specification, and load the resulting GUIDs into the array referred
+        // to by the specified 'result'.  Optionally specify 'numGuids',
+        // indicating the number of GUIDs to load into the 'result' array.  If
+        // 'numGuids' is not supplied, a default of 1 is used.  An RFC 4122
+        // version 4 GUID consists of 122 randomly generated bits, two
+        // 'variant' bits set to '10' and four 'version' bits set to '0100'.
+        // The behavior is undefined unless 'result' refers to a contiguous
+        // sequence of at least 'numGuids' Guid objects.  Note that this
+        // function uses the PCG random engine to generate high quality, albeit
+        // not cryptographically secure, random numbers for Guids.
```
Comment (mversche) [inline groups/bdl/bdlb/bdlb_guidutil.h]: We should explain why someone might chose t use this method:

...random numbers for Guids, which is typically *faster* than generating cryptographically secure guids.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+
+#include <bsl_cstdint.h>
+
+namespace BloombergLP {
+namespace bdlb {
+
+                                 // ========================
+                                 // class PCGRandomGenerator
+                                 // ========================
+class PCGRandomGenerator {
```
Comment (mversche) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: I think, based on the preponderance of similar abbreviations in BDE (e.g. `bdlt::Iso8601`, `bdlt::DateTz`, `bdlde::Md5`) that this should be `PcgRandomGenerator`

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+  public:
+    // CREATORS
+    PCGRandomGenerator(bsl::uint64_t initState      = 0,
+                       bsl::uint64_t streamSelector = 1);
+        // Create a 'PCGRandomGenerator' object and seed it with the optionally
+        // specified 'initState' and 'streamSelector.'  If 'initState' is not
+        // specified, 0 is used as the initial state.  If 'streamSelector' is
+        // not specified, 1 is used as the stream selector.
+
+    //! PCGRandomGenerator(const PCGRandomGenerator& original) = default;
```
Comment (mversche) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: Why is this a value semantic type?

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+
+    bsl::uint64_t d_streamSelector;  // selected sequence (stream)
+
+    // FRIENDS
+    friend bool operator==(const PCGRandomGenerator& lhs,
+                           const PCGRandomGenerator& rhs);
+
+  public:
+    // CREATORS
+    PCGRandomGenerator(bsl::uint64_t initState      = 0,
```
Comment (mversche) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: What are 'initState' and 'streamSelector'?  they are not explained what they do or control.  

Further, there needs to be some discussion in this component (probably in whatever usage example we choose to show first) how these values should typically be chosen.   This should probably show both how one might want to deterministically select them, and how one might want to generate them randomly from bdlb_RandomDevice (or whatever).  

In summary: explain what `initState` and `streamSelector` do in the function doc here, then have the function doc refer to a usage example showing how someone might use these values

**Review by mversche — APPROVED** [review]
I would lean towards having:

`bsl::uint32_t bdlb::Random::generatePcg(PcgRandomGenerator *);`

Which I realize is window dressing, but is where users will likely first look for a function providing random numbers.

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 19]
```diff
@@ -199,6 +199,20 @@ struct GuidUtil {
         // specification, consisting of 122 randomly generated bits, two
         // 'variant' bits set to '10' and four 'version' bits set to '0100'.
 
+    static void generateNonSecure(Guid *result, bsl::size_t numGuids = 1);
+    // Generate a sequence of GUIDs meeting the RFC 4122 version 4
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]: bde-format

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 20]
```diff
@@ -211,7 +225,16 @@ struct GuidUtil {
         // The behavior is undefined unless 'result' refers to a contiguous
         // sequence of at least 'numGuids' Guid objects.  Note that this
         // function uses the PCG random engine to generate high quality, albeit
-        // not cryptographically secure, random numbers for Guids.
+        // not cryptographically secure, random numbers for Guids. This method
+        // is typically faster than generating cryptographically secure Guids.
+
+    static Guid generateNonSecure();
+    // Generate and return a single GUID meeting the RFC 4122 version 4
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]:
```suggestion
        // Generate and return a single GUID meeting the RFC 4122 version 4
```

File: groups/bdl/bdlb/bdlb_guidutil.t.cpp [hunk 21]
```diff
@@ -295,6 +298,98 @@ int main(int argc, char *argv[])
 
     cout << "TEST " << __FILE__ << " CASE " << test << endl;;
     switch (test)  { case 0:
+      case 8: {
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.t.cpp:301]: Make this case 7, Usage - case 8.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 22]
```diff
@@ -49,81 +49,80 @@ BSLS_IDENT("$Id: $")
...
-//      } while (result > 5);
-//
-//      return result + 1;
-//  }
+
+// MANIPULATORS
+// int Die::roll()
+// {
+//     int result;
+
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: `//`

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 23]
```diff
@@ -49,81 +49,80 @@ BSLS_IDENT("$Id: $")
...
-//
-//  }
+// void shareStateAndSequence()
+// {
+//     Die a(123, 456);  // BAD IDEA
+//     Die b(123, 456);  // BAD IDEA
+
+//     int d1 = a.roll();
+//     int d2 = b.roll();
+//     ASSERT(d2 == d1);
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
//     assert(d2 == d1);
```

File: groups/bdl/bdlb/bdlb_random.h [hunk 24]
```diff
@@ -153,6 +155,10 @@ struct Random {
         // Return a 15-bit random number in the range '[ 0 .. 32,767 ]'
         // generated from the specified 'seed', and load into 'seed' a value
         // suitable for generating the next random number.
+
+    static bsl::uint32_t generatePcg(PcgRandomGenerator *generator);
+        // Return the next unsigned 32-bit random number generated from the
+        // specified PcgRandomGenerator.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_random.h]:
```suggestion
        // specified 'PcgRandomGenerator'.
```

File: groups/bdl/bdlb/bdlb_random.t.cpp [hunk 25]
```diff
@@ -323,31 +350,59 @@ int main(int argc, char *argv[])
                                   "==============" "\n";
 
         enum { NUM_ITERATIONS = 2000 };
+        {
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_random.t.cpp:384]: Add `generatePcg` to the test plan.

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 26]
```diff
@@ -124,18 +124,26 @@ Guid GuidUtil::generate()
     return result;
 }
 
-void GuidUtil::generate(unsigned char *result, bsl::size_t numGuids)
+namespace {
+inline
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp]:
```suggestion
```

File: groups/bdl/bdlb/bdlb_guidutil.cpp [hunk 27]
```diff
@@ -124,18 +124,26 @@ Guid GuidUtil::generate()
...
+void makeRfc4122Compliant(unsigned char *bytes, unsigned char *end)
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.cpp:117]: Move this function into the anonymous namespace at the beginning of the file:  we always put such implementation helper functions before implementation of any functions from the header.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 28]
```diff
@@ -153,11 +153,11 @@ namespace bdlb {
...
-    // This class implements a random number generator (RNG) based on
-    // the PCG algorithm.  PCG stands for "permuted congruential generator."
-    // The state is 64 bits.  It uses a so-called stream selector, also 64
-    // bits. For details, please refer below to the constructor.  For details of the algorithm,see
-    // http://www.pcg-random.org.
+    // This class implements a random number generator (RNG) based on the PCG
+    // algorithm.  PCG stands for "permuted congruential generator." The state
+    // is 64 bits.  It uses a so-called stream selector, also 64 bits. For
+    // details, please refer below to the constructor.  For details of the
+    // algorithm,see http://www.pcg-random.org.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
    // algorithm, see http://www.pcg-random.org.
```

**Review by khlebnikov — COMMENTED** [review]
Couple small comments.

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 29]
```diff
@@ -37,6 +37,14 @@ BSLS_IDENT("$Id: $")
...
 // <X>                ::=  [0123456789ABCDEFabcdef]{2}
 //
+///RANDOMNESS:
+///-----------
+// There are two different ways provided for generating GUIDs:
+// 1. 'generate': comparatively more secure but slower
+// 2. 'generateNonSecure': comparatively less secure but faster
+// 'generateNonSecure' employs the PCG random engine toproduce 
+// high-quality, but not cryptographically secure, GUIDs.
+//
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]: 1. I think the proper name for this section is `Cryptographic Security` rather than `RANDOMNESS`
2. `EXAMPLES` below refers to the `GUID String Format` section, so the security section should go after those examples.
3. We need to provide the people who read our doc with a coherent narrative, otherwise it's difficult to understand without prior knowledge.  Maybe something like this:

'GuidUtil' provides two families of functions for generating GUIDs: 'generate' and 'generateNonSecure'.  The primary difference between them is cryptographic security of the resulting GUIDs and performance. The slower 'generate' methods aim to produce cryptographically secure GUIDs by accessing the underlying system resources to obtain truly random numbers, whereas faster 'generateNonSecure' methods use a fast high-quality (but not strictly cryptographically secure) in-process random-number generator.

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 30]
```diff
@@ -210,8 +218,8 @@ struct GuidUtil {
         // The behavior is undefined unless 'result' refers to a contiguous
         // sequence of at least 'numGuids' Guid objects.  Note that this
         // function uses the PCG random engine to generate high quality, albeit
-        // not cryptographically secure, random numbers for Guids. This method
-        // is typically faster than generating cryptographically secure Guids.
+        // not cryptographically secure, random numbers for GUIDs. This method
+        // is typically faster than generating cryptographically secure GUIDs.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]: I think we should remove mention of our implementation details from out public documentation.  Maybe we decide to switch to something else than PCG - it shouldn't matter to our clients. Given our changes to the component documentation, I think we should only leave `Note that this function generates high-quality, albeit not cryptographically secure GUIDs.` Kind of "for safety" and to explain the function name.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 31]
```diff
@@ -16,7 +16,17 @@ BSLS_IDENT("$Id: $")
 // 'bdlb::PcgRandomGenerator', that is used to generate random numbers
 // employing the PCG algorithm, a high-performance, high-quality RNG.  The PCG
 // technique employs the concepts of permutation functions on tuples and a base
-// linear congruential generator.
+// linear congruential generator. In the PCG algorithm, rather than a single
+// seed parameter, there are two parameters, 'initState' and 'streamSelector'.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// linear congruential generator. The PCG algorithm is seeded with two 
// values: its initial state and a "stream selector".
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 32]
```diff
@@ -16,7 +16,17 @@ BSLS_IDENT("$Id: $")
...
+// The use of a second parameter ('streamSelector') is to address a potential
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// Stream selector is intended to address a potential
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 33]
```diff
@@ -16,7 +16,17 @@ BSLS_IDENT("$Id: $")
...
+// hazard when multiple instances of a random number generator are used: an
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// hazard of multiple instances of a random number generators having unintended correlation between their outputs.
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 34]
```diff
@@ -184,16 +194,7 @@ class PcgRandomGenerator {
...
-        // 'streamSelector') instead of a simple seed?  There is a potential
-        // hazard if multiple instances of a random number generator are used:
-        // an unintended correlation between their outputs.  For example, if we
-        // allow them to have the same internal state (e.g. mistakenly seeding
-        // both with the current time in seconds), they will output the exact
-        // same sequence of numbers.  Employing a stream selector enables the
-        // use of the same internal state.  Please refer to O’Neill (2014) at
-        // https://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf for further
-        // details of the algorithm.
+        // invocations of generate().
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
        // invocations of 'generate()'. 
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 35]
```diff
@@ -0,0 +1,307 @@
...
+//  }
+//  uint64_t streamSelector;
+//  if (0 != RandomDevice::getRandomBytes((unsigned char *)&streamSelector,
+//                                              sizeof(streamSelector))) {
+//           streamSelector = time(NULL) ^
+//                   (intptr_t)&bsl::printf;  // fallback streamSelector
+//  }
+//  bdlb::PcgRandomGenerator  rng(state, streamSelector);
+//  bsl::uint32_t randomInt = rng.generate();
+//..
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: We still need to replace/update this example.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 36]
```diff
@@ -0,0 +1,307 @@
...
+
+                                 // ========================
+                                 // class PcgRandomGenerator
+                                 // ========================
+class PcgRandomGenerator {
+    // This class implements a random number generator (RNG) based on the PCG
+    // algorithm.  PCG stands for "permuted congruential generator." The state
+    // is 64 bits.  It uses a so-called stream selector, also 64 bits. For
+    // details, please refer below to the constructor.  For details of the
+    // algorithm, see http://www.pcg-random.org.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: Incorporate this sentence `PCG stands for "permuted congruential generator."` and the link in component doc. Leave only the first sentence here.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 37]
```diff
@@ -0,0 +1,307 @@
...
+    // FRIENDS
+    friend bool operator==(const PcgRandomGenerator& lhs,
+                           const PcgRandomGenerator& rhs);
+
+  public:
+    // CREATORS
+    PcgRandomGenerator(bsl::uint64_t initState      = 0,
+                       bsl::uint64_t streamSelector = 1);
+        // Create a 'PcgRandomGenerator' object and seed it with the optionally
+        // specified 'initState' and 'streamSelector.'  If 'initState' is not
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
        // specified 'initState' and 'streamSelector'.  If 'initState' is not
```
The ticks are for code font, they don't abide by the rules of English text, so the period goes outside the quote.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 38]
```diff
@@ -0,0 +1,307 @@
...
+        // specified, 0 is used as the initial state.  If 'streamSelector' is
+        // not specified, 1 is used as the stream selector. 'initState' is the
+        // starting state for the RNG.  Any 64-bit value may be passed.
+        // 'streamSelector' selects the output sequence for the RNG.  Any
+        // 64-bit value may be passed, although only the low 63 bits are
+        // significant.  There are 2^63 different RNGs available, and
+        // 'streamSelector' selects from among them.  Invoking different
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: Too much information here, which is at the same time unclear. Most of it covered in component doc, so let's keep this simple.  

```suggestion
        // not specified, 1 is used as the stream selector.  Only the 
        // low 63 bits of 'streamSelector' are
        // significant.  Note that invoking different
```

File: groups/bdl/bdlb/bdlb_random.t.cpp [hunk 39]
```diff
@@ -217,7 +220,8 @@ int main(int argc, char *argv[])
       case 3: {
         // --------------------------------------------------------------------
         // USAGE EXAMPLE
-        //   Extracted from component header file.
+        //   Extracted from component header file. For 'generatePcg' (PCG)
+        //   usage example please refer bdlb_pcgrandomgenerator.t.cpp.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_random.t.cpp]:
```suggestion
        //   Extracted from component header file. 
```
This is not really relevant for the test driver.  The PCG stuff should be just be linked in component documentation in `@SEE_ALSO` section - search the code for where it goes and how it's formatted.

File: groups/bdl/bdlb/bdlb_random.h [hunk 40]
```diff
@@ -153,6 +155,10 @@ struct Random {
         // Return a 15-bit random number in the range '[ 0 .. 32,767 ]'
         // generated from the specified 'seed', and load into 'seed' a value
         // suitable for generating the next random number.
+
+    static bsl::uint32_t generatePcg(PcgRandomGenerator *generator);
+        // Return the next unsigned 32-bit random number generated from the
+        // specified 'PcgRandomGenerator'.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_random.h]:
```suggestion
        // specified 'generator'.
```

File: groups/bdl/bdlb/bdlb_guidutil.h [hunk 41]
```diff
@@ -217,9 +220,9 @@ struct GuidUtil {
         // 'variant' bits set to '10' and four 'version' bits set to '0100'.
         // The behavior is undefined unless 'result' refers to a contiguous
         // sequence of at least 'numGuids' Guid objects.  Note that this
-        // function uses the PCG random engine to generate high quality, albeit
-        // not cryptographically secure, random numbers for GUIDs. This method
-        // is typically faster than generating cryptographically secure GUIDs.
+        // function generates high quality, albeit not cryptographically
+        // secure, random numbers for GUIDs. This method is typically faster
+        // than generating cryptographically secure GUIDs.
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_guidutil.h]: Remove the last sentence - it doesn't refer to this function only, thus this information is in the component doc.  Also please update the rest of `generateNonSecure` function doc to match.

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 42]
```diff
@@ -129,29 +129,36 @@ BSLS_IDENT("$Id: $")
...
-// The PcgRandomGenerator constructor takes two 64-bit constants (the initial
-// state, and the rng sequence selector.  Instances of 'PcgRandomGenerator' with
-// different sequence selectors will never have random sequences that coincide.
-// In order to obtain nondeterministic output for each instance, on a Unix
-// system random bytes can be obtained from '/dev/urandom'.  The component 
-// bdlb_randomdevice provides such functionality.  Also illustrated below is a 
-// a backup mechanism for obtaining random bytes.
+///Example 2: Generating unique random sequences of 32-bit numbers
+///- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+// This example shows how one might construct two sequences of random numbers
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]: I took a stab at this example - please see the differences in the delivery style (also note that code is indented with 2 spaces after `//`):

```
///Example 2: Using stream selector to guarantee uncorrelated sequences
/// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// This example illustrates how the stream selector can be used to guarantee
// that two distinct instances of 'PCGRandomGenerator' produce uncorrelated
// sequences.
//
// First, we use the 'RandomDevice' to choose the initial states for the
// generators using a source of true randomness:
//..
//  uint64_t state1;
//  int rc = bdlb::RandomDevice::getRandomBytes(
//      reinterpret_cast<unsigned char *>(&state1), sizeof(state1));
//  (void)rc;  // error handling omitted
//
//  uint64_t state2;
//  int rc = bdlb::RandomDevice::getRandomBytes(
//      reinterpret_cast<unsigned char *>(&state2), sizeof(state2));
//  (void)rc;  // error handling omitted
//..
// Then we select two distinct stream selectors for the generators, which, due
// to the PCG algorithmic properties will guarantee that the sequences will be
// uncorrelated even if the initial state is exactly the same:
//..
//  const uint64_t streamSelector1 = 1;
//  const uint64_t streamSelector2 = 2;
//..
// Finally, we initialize the generators with their respective initial states
// and stream selectors and check that they produce distinct sequences of
// random numbers.  The check is guaranteed to pass even in the rare, but
// possible case of 'state1 == state2'.
//..
//  bdlb::PcgRandomGenerator rng1(state1, streamSelector1);
//  bdlb::PcgRandomGenerator rng2(state2, streamSelector2);
//
//  const int NUM_VALUES = 1000;
//  bsl::vector<bsl::uint32_t> sequence1(NUM_VALUES);
//  bsl::vector<bsl::uint32_t> sequence2(NUM_VALUES);
//
//  for (int i = 0; i < NUM_VALUES; ++i) {
//      sequence1[i] = rng1.generate();
//      sequence2[i] = rng2.generate();
//  }
//
//  assert(sequence1 != sequence2);
//..
```

File: groups/bdl/bdlb/bdlb_pcgrandomgenerator.h [hunk 43]
```diff
@@ -14,19 +14,19 @@ BSLS_IDENT("$Id: $")
...
-// unintended correlation between their outputs.  For example, if we allow them
-// to have the same internal state (e.g.  mistakenly seeding them with the
-// current time in seconds), they will output the exact same sequence of
-// numbers.  Employing a 'streamSelector' enables the the same 'initState' to
-// generate unique sequences.  Please refer to O’Neill (2014) at
-// https://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf for further details of
-// the algorithm.
+// employing the PCG algorithm, a high-performance, high-quality RNG.  PCG
+// stands for "permuted congruential generator." For details of the algorithm,
+// see http://www.pcg-random.org.  The PCG technique employs the concepts of
```
Comment (khlebnikov) [inline groups/bdl/bdlb/bdlb_pcgrandomgenerator.h]:
```suggestion
// stands for "permuted congruential generator" (see http://www.pcg-random.org 
// for details).  The PCG technique employs the concepts of
```
